bin/i18n-map-extractor.js

```diff
@@ -0,0 +1,145 @@
+const recast = require( 'recast' );
+const { forEach, reduce, includes, isEmpty, isFunction } = require( 'lodash' );
+const { readFileSync, writeFileSync } = require( 'fs' );
+const NormalModule = require( 'webpack/lib/NormalModule' );
+
+class wpi18nExtractor {
+	constructor( options ) {
+		const DEFAULT_FUNCTIONS = [
+			'__',
+			'_n',
+			'_x',
+			'_nx',
+			'sprintf',
+		];
+		this.options = options || {};
+		this.options.aliases = this.options.aliases || {};
+		this.options.filename = this.options.filename || 'translation-map.json';
+		this.translationMap = {};
+		this.functionNames = this.options.functionNames || DEFAULT_FUNCTIONS;
+	}
+
+	extractStringFromFunctionCall( args, functionName ) {
+		const strings = [];
+		switch ( functionName ) {
+			case '__':
+			case 'sprintf':
+			case '_n':
+				strings.push( args[ 0 ].value );
+				break;
+			case '_x':
+				strings.push( args[ 1 ].value + '\u0004' + args[ 0 ].value );
+				break;
+			case '_nx':
+				strings.push( args[ 3 ].value + '\u0004' + args[ 0 ].value );
+		}
+		return strings;
+	}
+
+	getStringsFromModule( module, extractor ) {
+		const source = module.originalSource().source();
+		if ( isEmpty( source ) ) {
+			return [];
+		}
+		let strings = [];
+		try {
+			const ast = recast.parse( source );
```

Comment:
> Is it necessary to create another variable to hold the results of this function call?

Are you referencing `source` being used for module.orginalSource().source()?  If so, short answer is no its not absolutely necessary but it means I'd be doing `module.originalSource().source()` call multiple times instead of just once.  Performance wise I don't think that's a good idea.

> Instead of declaring source as a const, can we not use let instead and then update its value at this step?

Sorry I don't understand why you think `let` is better than `const` here?  I intentionally used `const` because it _should not_ be changed within the scope of this method.  Can you elaborate on why you think `let` would be a better choice?

Reply:
> I intentionally used const because it should not be changed within the scope of this method. 

`source` isn't used anywhere else in the function other than to create `ast` so I didn't see why that would be the case

> Can you elaborate on why you think let would be a better choice?

the only reason I suggested is because it saves a variable and (imho) makes things more readable by reducing the necessity for that extra variable.

originally I was thinking that line 40 could use `let` instead of `const` and then you could just use `source` again in line 46 instead of the new `ast` variable.

but after looking at it a second time, line 40:

```js
let source = module.originalSource().source();
```

could be removed completely, which would require line 46 to change to:

```js
let source = recast.parse( module.originalSource().source() );
```

(which could use `const` instead of `let` if you really don't want that getting changed in any future iterations)
and then line 48 after this would change to:

```js
types.visit( source , {
```

in terms of whether `const` is better than `let` I don't think it matters as they are both block scoped and the only difference is that `const` is immutable. I was partially questioning why `source` needed to be immutable, which didn't make sense to me because the var was not used anywhere in the method, ie: why create an immutable constant to hold a value that you only reference once?

Reply:
This suggestion:

```
let source = recast.parse( module.originalSource().source() );
```

is a significant change of logic.  You may have missed the `isEmpty( source )` check in my code.  That's important because I'm skipping any processing when there is no source.  `recast.parse` throws an error  if source is empty.  Another reason for me doing things the way they are is to follow the "bail early" principle.  So I'm avoiding a lot of unnecessary logic if there is no original source value.

>  I don't think it matters as they are both block scoped and the only difference is that const is immutable. 

That's incorrect.  From MDN:

>  The const declaration creates a read-only reference to a value. It does not mean the value it holds is immutable, just that the variable identifier cannot be reassigned. For instance, in the case where the content is an object, this means the object's contents (e.g., its parameters) can be altered.

A pattern I seem to notice is (in order):

- favour use of `const` unless the variable needs reassigned at some point.
- favour use of `let` unless scope precludes using it.
- use `var`

> why create an immutable constant to hold a value that you only reference once?

Aside from the incorrect use of immutable.  The code under review is not referencing `source` once.  It's referenced in two places.  Once for an `isEmpty` check, and once with the recast call.